wcfsetup/install/files/lib/data/file/temporary/FileTemporary.class.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace wcf\data\file\temporary;
+
+use wcf\data\DatabaseObject;
+use wcf\util\JSON;
+
+/**
+ * @author Alexander Ebert
+ * @copyright 2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.1
+ *
+ * @property-read string $identifier
+ * @property-read int|null $time
+ * @property-read string $filename
+ * @property-read int $fileSize
+ * @property-read string $fileHash
+ * @property-read int|null $objectTypeID
+ * @property-read string $context
+ * @property-read string $chunks
+ */
+class FileTemporary extends DatabaseObject
+{
+    protected static $databaseTableIndexIsIdentity = false;
+
+    protected static $databaseTableIndexName = 'identifier';
+
+    public const MAX_CHUNK_COUNT = 255;
+
+    public function getChunkCount(): int
+    {
+        return \strlen($this->chunks);
+    }
+
+    public function getChunkSize(): int
+    {
+        return \ceil($this->fileSize / $this->getChunkCount());
+    }
+
+    public function hasChunk(int $sequenceNo): bool
+    {
+        return $this->chunks[$sequenceNo] === '1';
+    }
+
+    public function getFilename(): string
+    {
+        return \sprintf("%s.bin", $this->identifier);
+    }
+
+    public function getPath(): string
+    {
+        $folderA = \substr($this->identifier, 0, 2);
+        $folderB = \substr($this->identifier, 2, 2);
+
+        return \sprintf(
+            \WCF_DIR . '_data/private/fileUpload/%s/%s/',
+            $folderA,
+            $folderB,
+        );
+    }
+
+    public function getPathname(): string
+    {
+        return $this->getPath() . $this->getFilename();
+    }
+
+    public function getContext(): array
+    {
+        return JSON::decode($this->context);
+    }
+
+    public static function getNumberOfChunks(int $fileSize): int
+    {
+        return \ceil($fileSize / self::getOptimalChunkSize());
+    }
```

Comment:
The result of this function will change if `post_max_size` changes. Will this cause problems?

Reply:
Yes, but only if the value decreases while the chunks are being uploaded. There is nothing we can do about this because the client needs to know what the acceptable values are.

Reply:
Okay, if it's only during the upload, then it's probably acceptable. However I suggest to check the failure mode at least: Will the upload appear stuck to the user? Will it “gracefully” report an error message that the upload needs to be restarted? Will it result in a corrupt file, without any indication of an error?

Reply:
The upload will fail in this case because the checksum will mismatch which triggers the error handler. This is the beauty of strict checksum checks for both the individual chunks and the complete file, guarding against corrupted files.